quaddtype/numpy_quaddtype/src/casts.cpp

```diff
@@ -204,91 +320,162 @@ numpy_to_quad_strided_loop(PyArrayMethod_Context *context, char *const data[],
+    if (backend == BACKEND_SLEEF) {
+        return (npy_int)Sleef_cast_to_int64q1(x.sleef_value);
+    }
+    else {
+        return (npy_int)x.longdouble_value;
+    }
 }
+
 template <>
 inline npy_uint
-from_quad<npy_uint>(Sleef_quad x)
+from_quad<npy_uint>(quad_value x, QuadBackendType backend)
 {
-    return (npy_uint)Sleef_cast_to_uint64q1(x);
+    if (backend == BACKEND_SLEEF) {
+        return (npy_uint)Sleef_cast_to_uint64q1(x.sleef_value);
+    }
+    else {
+        return (npy_uint)x.longdouble_value;
+    }
 }
+
 template <>
 inline npy_long
-from_quad<npy_long>(Sleef_quad x)
+from_quad<npy_long>(quad_value x, QuadBackendType backend)
 {
-    return (npy_long)Sleef_cast_to_int64q1(x);
+    if (backend == BACKEND_SLEEF) {
+        return (npy_long)Sleef_cast_to_int64q1(x.sleef_value);
+    }
+    else {
+        return (npy_long)x.longdouble_value;
+    }
 }
+
 template <>
 inline npy_ulong
-from_quad<npy_ulong>(Sleef_quad x)
+from_quad<npy_ulong>(quad_value x, QuadBackendType backend)
 {
-    return (npy_ulong)Sleef_cast_to_uint64q1(x);
+    if (backend == BACKEND_SLEEF) {
+        return (npy_ulong)Sleef_cast_to_uint64q1(x.sleef_value);
+    }
+    else {
+        return (npy_ulong)x.longdouble_value;
+    }
 }
+
 template <>
 inline npy_longlong
-from_quad<npy_longlong>(Sleef_quad x)
+from_quad<npy_longlong>(quad_value x, QuadBackendType backend)
 {
-    return Sleef_cast_to_int64q1(x);
+    if (backend == BACKEND_SLEEF) {
+        return Sleef_cast_to_int64q1(x.sleef_value);
+    }
+    else {
+        return (npy_longlong)x.longdouble_value;
+    }
 }
+
 template <>
 inline npy_ulonglong
-from_quad<npy_ulonglong>(Sleef_quad x)
+from_quad<npy_ulonglong>(quad_value x, QuadBackendType backend)
 {
-    return Sleef_cast_to_uint64q1(x);
+    if (backend == BACKEND_SLEEF) {
+        return Sleef_cast_to_uint64q1(x.sleef_value);
+    }
+    else {
+        return (npy_ulonglong)x.longdouble_value;
+    }
 }
+
 template <>
 inline float
-from_quad<float>(Sleef_quad x)
+from_quad<float>(quad_value x, QuadBackendType backend)
 {
-    return Sleef_cast_to_doubleq1(x);
+    if (backend == BACKEND_SLEEF) {
+        return (float)Sleef_cast_to_doubleq1(x.sleef_value);
+    }
+    else {
+        return (float)x.longdouble_value;
+    }
 }
+
 template <>
 inline double
-from_quad<double>(Sleef_quad x)
+from_quad<double>(quad_value x, QuadBackendType backend)
 {
-    return Sleef_cast_to_doubleq1(x);
+    if (backend == BACKEND_SLEEF) {
+        return Sleef_cast_to_doubleq1(x.sleef_value);
+    }
+    else {
+        return (double)x.longdouble_value;
+    }
 }
+
 template <>
 inline long double
-from_quad<long double>(Sleef_quad x)
+from_quad<long double>(quad_value x, QuadBackendType backend)
 {
-    return Sleef_cast_to_doubleq1(x);
+    if (backend == BACKEND_SLEEF) {
+        return (long double)Sleef_cast_to_doubleq1(x.sleef_value);
+    }
+    else {
+        return x.longdouble_value;
+    }
 }
```

Comment:
I think it would be simpler to template over backend and output type, then you wouldn't need the runtime branching.